content/blog/2025-04-24-super-optimization/index.md

```diff
@@ -0,0 +1,104 @@
++++
+title = "Superoptimization: a quest for -O∞"
+[[extra.authors]]
+name = "Mark Barbone"
+[[extra.authors]]
+name = "Zihan Li"
+[[extra.authors]]
+name = "Samuel Breckenridge"
++++
+
+The -O flag makes your program faster. -O2 makes it faster still, and ideally, -O3 squeezes out even more speed, and the cost of being slower to compile. What if we could keep going? -O4, -O5, or even -O∞, getting the fastest program ever (at the expense of perhaps the slowest compiler)? This is the goal of superoptimization. In this blog post we’ll discuss the paper “Denali: A Goal-directed Superoptimizer”.
+
+## Background
+
+Although in general compilers “optimize” code while translating it to machine language, in most cases the resulting machine language will not be optimal in the sense of calculating the desired result in the fewest possible cycles. However, for some critical paths in some programs for which a programmer truly desires the best possible performance, this notion of true optimality might be appropriate. In these cases, one option is to resort to hand-written machine code. Clearly for hand-written code to approach optimality the programmer requires expert knowledge of the underlying architecture, and time to devote to iterations of optimizing and testing. If we could rely on automatic code generation to produce truly optimal code, we would be able to save expert programmer time and costs both when writing new code and when porting highly optimized code between different architectures. Denali, as a “superoptimizer”, aims to do exactly this.
+
+Two tools that Denali uses to do superoptimization, for which some background is helpful, are e-graphs and satisfiability solvers.
+
+### E-graphs 
+
+An e-graph is a DAG that represents the computation of a program. Its key feature is that it allows for equivalence relations to be specified between nodes in the graph so that it may represent many equivalent versions of the same program compactly. Today, there’s been a resurgence of interest from the PL research community in this idea, in part due to the very nice [egg](https://egraphs-good.github.io/) library for e-graphs.
+
+### Boolean Satisfiability (SAT) Solvers
+
+Denali was one of the earliest applications of SAT solvers to code generation, an approach which has since become widely adopted in the field (for example used in Sketch and Brahma). At the time the Denali paper was written, SAT solvers were improving at an insane rate: for at least 10-15 years, SAT algorithms were getting faster at about the same rate as Moore’s law.  Today’s SAT solvers solve problems with ~tens of millions of variables.
```

Comment:
How about hyperlinks to Sketch and Brahma as well?

Also, I'm not sure it matters that Sketch and Brahma both use SMT with plenty of theories while Denali used plain ol' SAT. Perhaps that is too low-level to matter, but it might rankle readers who know something about the modern era of program synthesis.